Update to the JDBCStore section of the Transaction guide (#36179)
quarkusio/quarkus: docs/src/main/asciidoc/transaction.adoc

and pull requests should be submitted there:
https://github.com/quarkusio/quarkus/tree/main/docs/src/main/asciidoc
////
= Using transactions in Quarkus
include::_attributes.adoc[]
:diataxis-type: reference
:categories: data,getting-started
:topics: data,jpa,jta,transactions,narayana
:extensions: io.quarkus:quarkus-narayana-jta

> Quarkus includes a Transaction Manager extension used to coordinate and expose transactions to your applications as described in the link: https://jakarta.ee/specifications/transactions/[Jakarta Transactions] specification, formerly known as Java Transaction API (JTA).

Reviewer: May I ask why you added "extension" to "Transaction Manager"? Could you reference it?

Reviewer 2: I tend to agree with @jmfinelli this is not the best wording, since there is no Quarkus extension named even remotely like that (the extension is called `quarkus-narayana-jta`, for better or worse). Plus "Transaction Manager extension" can be interpreted as "something that extends a transaction manager", which would be incorrect.

If you want to mention that extensions are involved, I guess this could work:

```suggestion
The `quarkus-narayana-jta` extension provides a Transaction Manager that coordinates and expose transactions to your applications as described in the link: https://jakarta.ee/specifications/transactions/[Jakarta Transactions] specification, formerly known as Java Transaction API (JTA).
```

When discussing Quarkus transactions, this guide refers to Jakarta Transactions transaction style and uses only the term _transaction_ to address them.

Also, Quarkus does not support distributed transactions.
This means that models that propagate transaction context, such as link:https://download.oracle.com/otndocs/jcp/7309-jts-1.0-spec-oth-JSpec/[Java Transaction Service] (JTS), REST-AT, WS-Atomic Transaction, and others, are not supported by the `narayana-jta` extension.

> The  `narayana-jta` extension supports only the Jakarta Transactions model but does not support its JTS component, which is used for interoperability between Transaction Managers.

Reviewer 3: not sure if it matters in AsciiDoc, but there are 2 spaces between `the` and `narayana-jta`

> The `narayana-jta` extension supports only the Jakarta Transactions model but does not support its JTS component, which is used for interoperability between Transaction Managers.

Reviewer: Could you please reference where JTS is defined as a "component" of the Jakarta Transactions specification? Thanks.

Reviewer 2: It's not a "component", it's a separate spec.

Anyway, I think we can remove this line as the paragraph above already states "models [...] such as Java Transaction Service (JTS) [...] are not supported by the `narayana-jta` extension.

Agreed @MichalMaler ?

////
= Using Transactions in Quarkus
include::_attributes.adoc[]
:categories: data,transactions

Quarkus includes a Transaction Manager extension used to coordinate and expose transactions as described in the link: https://jakarta.ee/specifications/transactions/[Jakarta Transactions] specification, formerly known as Java Transaction API (JTA), to your applications.

When discussing Quarkus transactions, this guide refers to Jakarta Transactions transaction style and thereafter uses the term _transaction_ to address them.

> NOTE: Quarkus does not support distributed transactions.

Reviewer: Could you please expand this sentence? Michael's [comment](https://github.com/quarkusio/quarkus/pull/36179#issuecomment-1750293436) could be used to clarify what "distributed transactions" means here.

////
= Using Transactions in Quarkus
include::_attributes.adoc[]
:categories: data,transactions

Quarkus includes a Transaction Manager extension used to coordinate and expose transactions to your applications as described in the link: https://jakarta.ee/specifications/transactions/[Jakarta Transactions] specification, formerly known as Java Transaction API (JTA).

When discussing Quarkus transactions, this guide refers to Jakarta Transactions transaction style and uses only the term _transaction_ to address them.

Also, Quarkus does not support distributed transactions.

> This means that models that propagate transaction context, such as JTS Topology Suite (JTS), JBoss Remoting, REST-AT, XTS, and others, are not supported by the `narayana-jta` extension.

Reviewer: My bad, I should have specified that JTS stands for [Java Transaction Service](https://download.oracle.com/otndocs/jcp/7309-jts-1.0-spec-oth-JSpec/)

Reviewer: Can we remove `JBoss Remoting` from the list? We should also replace XTS with `WS-Atomic Transaction`.

> The  `narayana-jta` extension supports only the Jakarta Transactions model but does not support its JTS component, which is used for interoperability between Transaction Managers.

Reviewer: Isn't this redundant?
BTW, I am not sure we can call JTS a component of the Takarta Transactions spec

Reviewer: Could you please reference where JTS is called "component" in the JTA spec?

[[jdbcstore]]
== Configure storing of Quarkus transaction logs in a database

In cloud environments where persistent storage is not available, such as when application containers are unable to use persistent volumes, you can configure the transaction management to store transaction logs in a database by using a Java Database Connectivity (JDBC) datasource.

> NOTE: This configuration is only relevant for _Jakarta Transactions_ transactions.

Reviewer: Shouldn't we use only _transactions_ here?

In cloud environments where persistent storage is not available, such as when application containers are unable to use persistent volumes, you can configure the transaction management to store transaction logs in a database by using a JDBC datasource.

NOTE: This configuration is only relevant for _Jakarta Transactions_ transactions.

> However, in cloud-native apps, it is important to assess JTA transactions after careful evaluation.

Reviewer 4: The plan was to define what transaction model the extension supports in the introductory paragraph and thereafter to just use the term transaction. I think Manuel also highlighted the use of `JTA` in one of his earlier comments.

> The Narayana framework, which manages these transactions, requires stable storage, a unique reusable node identifier, and a steady IP address to work correctly.

Reviewer 4: I'd stick to the name of the extension rather than using terms like `Narayana framework`.

Reviewer: +1

> While the JDBC object store handles one of these needs, users must still plan how to meet the other two requirements.

Reviewer 4: Which one?

Reviewer: ```suggestion
While the JDBC object store provides a stable storage, users must still plan how to meet the other two requirements.
```

Quarkus then allows the following JDBC-specific configuration of the object store included in `quarkus.transaction-manager.object-store._<property>_` properties, where _<property>_ can be:

> * `type` (_string_): Configure this property to `jdbc` to enable usage of a Quarkus JDBC datasource for transaction logging.

Reviewer 4: I assume "transaction logging" is referring to transaction logs as opposed to log messages. To be consistent with the rest of the document I would write `storing transaction logs` instead of "for transaction logging".

In cloud environments where persistent storage is not available, such as when application containers are unable to use persistent volumes, you can configure the transaction management to store transaction logs in a database by using a JDBC datasource.

[IMPORTANT]
====

> Using a database to store transaction logs has several benefits, but in comparison with using the file system to store the logs, you might notice a reduction in performance.

Reviewer: s/benefits, but in comparison/benefits but, in comparison/

Reviewer 4: If you are going to recommend one over the other you need delineate the pros and cons of each store.

Storing transaction logs into a database is slower, creates bigger overhead, and requires higher maintenance.

> Use this approach when prioritizing consistency over throughput.

Reviewer: I don't think this sentence is true. Consistency must always be guaranteed in XA transactions.

> Use this approach when prioritizing consistency over throughput.

Reviewer: I don't think this sentence is true. Consistency must always be guaranteed in XA transactions. Could you please remove this sentence from the PR? Thanks

Reviewer 4: +1 saying that it is slow, adds overhead and is higher maintenance depends upon how the environment is configured so it is not a true statement. If you want to provide a cost benefit analysis of the two stores you must proved a thorough balanced view of the two approaches and indicate the circumstances under which one type shines over the other. Can you also indicate under which circumstances consistency is compromised, a user would need to know, otherwise he will in all likelihood read this and opt for a transaction manager that does provide JTA semantics.

> In addition, this methodology is only relevant for eXtended Architecture (XA) scenarios.

Reviewer: Two things here:
* Could we replace 'methodology' with 'configuration'?
* Could we replace 'eXtended Architecture (XA) scenarios' with 'eXtended Architecture (XA) transactions' or 'distributed (XA) transactions'.

> In addition, this configuration is only relevant for eXtended Architecture (XA) transactions.

Reviewer 4: I'd stick to the standard terminology: in the DTP standards (https://pubs.opengroup.org/onlinepubs/9294999599/toc.pdf and htps://pubs.opengroup.org/onlinepubs/009680699/toc.pdf) there is no reference to eXtended Architecture and I don't think it is de facto either.

> However, use XA in cloud-native applications only if explicitly required by your use case.

Reviewer: Two things here:
* 'XA' alone doesn't sound clear to me...'XA transactions' is better
* 'only if explicitly required by your use case' could be replaced with 'only after careful evaluations'

Probably, we would like to add a sentence here to be more specific why XA transactions need 'careful evaluations' in cloud-native applications. In particular, Narayana (i.e. the framework that handles XA transactions) needs a stable storage, a unique (and reusable) node identifier and a stable IP to run XA transactions. When the JDBC object store is used, one of these requirements is fulfilled but users still need to consider strategies to satisfy the other two requirements. More details [here](https://jbossts.blogspot.com/2022/04/narayana-on-cloud-part-1.html)

> However, use XA transactions in cloud-native applications only if explicitly required by your use case.

Reviewer 4: Why do you use the term XA rather than JTA?

In cloud environments where persistent storage is not available, such as when application containers are unable to use persistent volumes, you can configure the transaction management to store transaction logs in a database by using a JDBC datasource.

[IMPORTANT]
====
Using a database to store transaction logs has several benefits, but in comparison with using the file system to store the logs, you might notice a reduction in performance.

> In addition, this configuration is only relevant for JTA transactions.

Reviewer: Thanks for replacing XA with JTA. Re-reading this, I was wondering if we should use `Jakarta Transactions transactions`

Reviewer: This is valid for all JTA occurrences

Reviewer 4: The very first paragraph in [this guide](https://quarkus.io/guides/transaction) should say that the extension supports Jakarta Transactions transactions style of transaction. And thereafter the guide should only use the term transaction. In this way you can avoid the repetitive and clunky use of the phrase `Jakarta Transactions transaction`.

Reviewer: +1

If no value is provided for the `datasource` property, Quarkus uses the xref:datasource.adoc#configure-datasources[default datasource].

* `create-table` (_boolean_): When set to `true`, the transaction log table gets automatically created if it does not already exist.
The default value is `false`.

* `drop-table` (_boolean_): When set to `true`, the tables are dropped on startup if they already exist.
The default value is `false`.

* `table-prefix` (string): Specify the prefix for a related table name.
The default value is `quarkus_`.

For more configuration information, see the *Narayana JTA - Transaction manager* section of the Quarkus xref:all-config.adoc[All configuration options] reference.

.Additional information:

* You can manually create the transaction log table during the initial setup by setting the `create-table` property to `true`.

* JDBC datasources and ActiveMQ Artemis allow the enlistment and automatically register the `XAResourceRecovery`.

** JDBC datasources is part of `quarkus-agroal`, and it needs to use `quarkus.datasource.jdbc.transactions=XA`.
** ActiveMQ Artemis is part of `quarkus-pooled-jms`, and it needs to use `quarkus.pooled-jms.transaction=XA`.

> * To ensure data protection in case of application crashes or failures, enable the transaction crash recovery with the `quarkus.transaction-manager.enable-recovery=true` configuration.

Reviewer: We should replace `data protection` with `data integrity`

[[jdbcstore]]
== Configure storing of Quarkus transaction logs in a database

In cloud environments where persistent storage is not available, such as when application containers are unable to use persistent volumes, you can configure the transaction management to store transaction logs in a database by using a Java Database Connectivity (JDBC) datasource.

> However, in cloud-native apps, using an in-database transaction store has additional requirements.

Reviewer: Thanks @MichalMaler for modifying the `using a database to store transaction logs` at line 378. Could you please replace `using an in-database transaction store` with `using a database to store transaction logs` here as well. Thanks

Reviewer 2: It's done, thanks @jmfinelli .

BTW I agree the GitHub UI tends to be confusing when history gets rewritten, but generally I manage by assuming no unrelated changes get introduced between reviews, and simply going through all previous conversations on the "Conversations" tab. Having the "Files changed" tab opened in a different browser tab helps :octopus:

Reviewer: Thanks @yrodiere and @MichalMaler.

BTW, I see your point and thanks for the suggestion. Personally, I find long conversations hard/tricky to follow when comments disappear from "Files changed". For this reason, I usually rewrite history when all reviewers approved my PR.

[[jdbcstore]]
== Configure storing of Quarkus transaction logs in a database

In cloud environments where persistent storage is not available, such as when application containers are unable to use persistent volumes, you can configure the transaction management to store transaction logs in a database by using a Java Database Connectivity (JDBC) datasource.

> However, in cloud-native apps, it is important to assess transactions after careful evaluation.

Reviewer: I don't think this sentence is clear. What do you want to convey here?

Reviewer 2: New sentence from @MichalMaler LGTM:

> However, in cloud-native apps, using an in-database transaction store has additional requirements.

All good for you @jmfinelli ?

The `narayana-jta` extension, which manages these transactions, requires stable storage, a unique reusable node identifier, and a steady IP address to work correctly.
While the JDBC object store provides a stable storage, users must still plan how to meet the other two requirements.

> Quarkus then allows the following JDBC-specific configuration of the object store included in `quarkus.transaction-manager.object-store._<property>_` properties, where _<property>_ can be:

Reviewer: Removing the `IMPORTANT` keyword at (the old) line number 372, it seems to me that this sentence here becomes detached from the flow of the document. In fact, I didn't follow what `then` was referring to.

Reviewer 2: I guess "then" is meant as "after you assessed whether an in-database transaction store is right for you", @MichalMaler ?

It's probably something that would solve itself once we resolve the conversation above about "assess transactions after careful evaluation".

Reviewer 2: New wording LGTM:

> Quarkus, after you evaluated whether an in-database transaction store is right for you, allows the following JDBC-specific configuration of the object store included in `quarkus.transaction-manager.object-store._<property>_` properties, where _`<property>`_ can be:

> Quarkus, after you evaluated whether an in-database transaction store is right for you, allows the following JDBC-specific configuration of the object store included in `quarkus.transaction-manager.object-store._<property>_` properties, where _<property>_ can be:

Reviewer: As commits are constantly re-written, I cannot track when this was added and why...or if we already discussed it. Is the added sentence needed? In that case why? I'm asking as I would not recommend to use `in-database transaction store`. `using a database to store transaction logs` was way clearer.

* `type` (_string_): Configure this property to `jdbc` to enable usage of a Quarkus JDBC datasource for storing transaction logs.
The default value is `file-system`.

* `datasource` (_string_): Specify the name of the datasource for the transaction log storage.
If no value is provided for the `datasource` property, Quarkus uses the xref:datasource.adoc#configure-datasources[default datasource].

* `create-table` (_boolean_): When set to `true`, the transaction log table gets automatically created if it does not already exist.
The default value is `false`.

* `drop-table` (_boolean_): When set to `true`, the tables are dropped on startup if they already exist.
The default value is `false`.

* `table-prefix` (string): Specify the prefix for a related table name.
The default value is `quarkus_`.

// This paragraph will differ from the RHBQ docs in the future since it is not supported in the product yet.
Additional information:
For more configuration information, see the *Narayana JTA - Transaction manager* section of the Quarkus xref:all-config.adoc[All configuration options] reference.

> * You can manually create the transaction log table during the initial setup by setting the `create-table` property to `true`.

Reviewer: Why did you remove "manually"? Could you reference it?

Reviewer 3: @jmfinelli I asked for it. using [quarkus.transaction-manager.object-store.create-table](https://quarkus.io/guides/all-config#quarkus-narayana-jta_quarkus.transaction-manager.object-store.create-table) is a straight opposite to manually, it creates table **automatically**

Reviewer 2: Makes sense to me.

Reviewer: Thanks @MichalMaler.

* `table-prefix` (string): Specify the prefix for a related table name.
The default value is `quarkus_`.

For more configuration information, see the *Narayana JTA - Transaction manager* section of the Quarkus xref:all-config.adoc[All configuration options] reference.

.Additional information:

* You can manually create the transaction log table during the initial setup by setting the `create-table` property to `true`.

* JDBC datasources and ActiveMQ Artemis allow the enlistment and automatically register the `XAResourceRecovery`.

** JDBC datasources is part of `quarkus-agroal`, and it needs to use `quarkus.datasource.jdbc.transactions=XA`.
** ActiveMQ Artemis is part of `quarkus-pooled-jms`, and it needs to use `quarkus.pooled-jms.transaction=XA`.

> * To ensure data protection in case of application crashes or failures, enable the transaction crash recovery feature with the `quarkus.transaction-manager.enable-recovery=true` configuration, noting that the default value is `false`.

Reviewer 4: Describing transaction recovery as a `feature` and defaulting to `false` makes me laugh.

* `table-prefix` (string): Specify the prefix for a related table name.
The default value is `quarkus_`.

For more configuration information, see the *Narayana JTA - Transaction manager* section of the Quarkus xref:all-config.adoc[All configuration options] reference.

// * xref:all-config.adoc#uarkus-narayana-jta-narayana-jta-transaction-manager[Narayana JTA - Transaction manager].

.Additional information:

* You can manually create the transaction log table during the initial setup by setting the `create-table` property to `true`.

* JDBC datasources and ActiveMQ Artemis allow the enlistment and automatically register the `XAResourceRecovery`.

** JDBC datasources is part of `quarkus-agroal`, and it needs to use `quarkus.datasource.jdbc.transactions=XA`.
** ActiveMQ Artemis is part of `quarkus-pooled-jms`, and it needs to use `quarkus.pooled-jms.transaction=XA`.

> * If your application employs XA transactions, enable the transaction crash recovery feature by using `quarkus.transaction-manager.enable-recovery=true`, to safeguard your data in the event of application crashes or failures.

Reviewer 4: Why does the documentation keep referring to XA transactions, the extension is called the `narayana-jta` extension because it supports JTA style transactions, mixing terminology creates a barrier to understanding and will only adds confusion.

> The default value for XA Recovery is `false`.

Reviewer: Why this info has been moved to a new line?

* `table-prefix` (string): Specify the prefix for a related table name.
The default value is `quarkus_`.

For more configuration information, see the *Narayana JTA - Transaction manager* section of the Quarkus xref:all-config.adoc[All configuration options] reference.

.Additional information:

* You can manually create the transaction log table during the initial setup by setting the `create-table` property to `true`.

* JDBC datasources and ActiveMQ Artemis allow the enlistment and automatically register the `XAResourceRecovery`.

** JDBC datasources is part of `quarkus-agroal`, and it needs to use `quarkus.datasource.jdbc.transactions=XA`.
** ActiveMQ Artemis is part of `quarkus-pooled-jms`, and it needs to use `quarkus.pooled-jms.transaction=XA`.

* If your application employs JTA transactions style, enable the transaction crash recovery feature by using `quarkus.transaction-manager.enable-recovery=true`, to safeguard your data in the event of application crashes or failures.

> The default value for JTA recovery is `false`.

Reviewer: Although the rendering of the document appears to add the sentence `The default value for JTA recovery is false.` on the same line...I'd like to suggest to put `The default value for JTA recovery is false.` on the same line.
```suggestion
* If your application employs JTA transactions style, enable the transaction crash recovery feature by using `quarkus.transaction-manager.enable-recovery=true`, to safeguard your data in the event of application crashes or failures. The default value for JTA recovery is `false`.
```

Reviewer 4: It is already clear that the guide is for JTA transactions, after all it is its raison d'etre. So just use the term `recovery` since JTA is implied - also see my earlier comment about the terminology around transactions.
